server/src/test/java/org/elasticsearch/cluster/coordination/DeterministicTaskQueueTests.java

```diff
@@ -0,0 +1,280 @@
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cluster.coordination;
+
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+
+import static org.elasticsearch.node.Node.NODE_NAME_SETTING;
+import static org.hamcrest.Matchers.contains;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.empty;
+import static org.hamcrest.core.Is.is;
+
+public class DeterministicTaskQueueTests extends ESTestCase {
+
+    public void testRunNextTask() {
+        final DeterministicTaskQueue taskQueue = newTaskQueue();
+        final List<String> strings = new ArrayList<>(2);
+
+        taskQueue.scheduleNow(() -> strings.add("foo"));
+        taskQueue.scheduleNow(() -> strings.add("bar"));
+
+        assertThat(strings, empty());
+
+        assertTrue(taskQueue.hasRunnableTasks());
+        taskQueue.runNextTask();
+        assertThat(strings, contains("foo"));
+
+        assertTrue(taskQueue.hasRunnableTasks());
+        taskQueue.runNextTask();
+        assertThat(strings, contains("foo", "bar"));
+
+        assertFalse(taskQueue.hasRunnableTasks());
+    }
+
+    public void testRunRandomTask() {
+        final DeterministicTaskQueue taskQueue = newTaskQueue();
+        final List<String> strings = new ArrayList<>(4);
+
+        taskQueue.scheduleNow(() -> strings.add("foo"));
+        taskQueue.scheduleNow(() -> strings.add("bar"));
+        taskQueue.scheduleNow(() -> strings.add("baz"));
+        taskQueue.scheduleNow(() -> strings.add("quux"));
+
+        assertThat(strings, empty());
+
+        while (taskQueue.hasRunnableTasks()) {
+            taskQueue.runRandomTask(random());
+        }
+
+        assertThat(strings, containsInAnyOrder("foo", "bar", "baz", "quux"));
+    }
+
+    public void testStartsAtTimeZero() {
+        final DeterministicTaskQueue taskQueue = newTaskQueue();
+        assertThat(taskQueue.getCurrentTimeMillis(), is(0L));
+    }
+
+    public void testDoesNotDeferTasksForImmediateExecution() {
+        final DeterministicTaskQueue taskQueue = newTaskQueue();
+        final List<String> strings = new ArrayList<>(1);
+
+        taskQueue.scheduleAt(0, () -> strings.add("foo"));
```

Comment:
instead of using the fixed value 0 here, can you first move forward the current time and then schedule something at that exact time?